src/libasr/pass/intrinsic_functions.h

```diff
@@ -3770,6 +3771,16 @@ namespace BitSize {
 
 } // namespace BitSize
 
+namespace SameTypeAs {
+
+    static ASR::expr_t *eval_SameTypeAs(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+        bool same_type = ASRUtils::expr_type(args[0]) == ASRUtils::expr_type(args[1]);
```

Comment:
I think this implementation can be improved. In the current implementation we only handle the base type. Also, as the return type of `ASRUtils::expr_type()` is a pointer of type `ASR::ttype_t*`, it may cause issues.

We can improve this by simply having a call to the `ASRUtils::check_equal_type()` utlity function.
```suggestion
        bool same_type = ASRUtils::check_equal_type(ASRUtils::expr_type(args[0]), ASRUtils::expr_type(args[1]));
```

Reply:
This didn't work for me. I'm still getting wrong output only.

Reply:
Hmm...can you try using the `ASRUtils::types_equal()` function? `ASRUtils::check_equal_types()` has a call to this function only, but still we can try it explicitly to see what happens.